R/tags.R

```diff
@@ -1182,14 +1204,37 @@ withTags <- function(code) {
 
 # Make sure any objects in the tree that can be converted to tags, have been
 tagify <- function(x) {
-  rewriteTags(x, function(uiObj) {
-    if (isResolvedTag(uiObj) || isTagList(uiObj) || is.character(uiObj))
-      return(uiObj)
-    else
-      tagify(as.tags(uiObj))
+
+  # Schedule .postRenderHook(s) to run _after_ converting to tags
+  postHooks <- list()
+  on.exit({
+    for (hook in postHooks) tryWarn(hook())
+  })
+
+  rewriteTags(x, function(ui) {
+    if (is.character(ui)) return(ui)
+
+    pre <- attr(ui, "renderHooks")
+    for (hook in pre) {
+      tryWarn(ui <- as.tags(hook(ui)))
+    }
+    attr(ui, "renderHooks") <- NULL
+
+    post <- attr(ui, "postRenderHooks")
+    if (length(post)) {
+      postHooks <<- c(postHooks, post)
+      attr(ui, "postRenderHooks") <- NULL
+    }
+
+    if (isTag(ui) || isTagList(ui)) ui else tagify(as.tags(ui))
+
```

Comment:
A couple things feel strange about this. As mentioned elsewhere, this doesn't use the return value from the post hook, but I think that should be possible.

One thing that feels strange is the asymmetry between how the pre and post hooks are run. The pre hooks are run from inside the anonymous function passed to `rewriteTags()`; in contrast, the post hooks are *collected* within the anonymous function, but they're *executed* by `tagify`'s `on.exit()`. Could the call to `on.exit()` be moved into the inner anonymous function?

Or could the invocation of both the pre and post render hooks happen outside of the anonymous function?

Reply:
> Could the call to on.exit() be moved into the inner anonymous function?

I don't think that's quite right since in that case the post hooks could run too early in some cases.